File: native/strapiclient/src/model/cat.rs
use crate::model::GalleryImage;
use rustler::NifStruct;

use crate::primitives::DateTime;

#[derive(Debug, NifStruct)]
#[module = "Kotkowo.Cat"]
pub struct Cat {
    pub id: usize,
    pub slug: String,
    pub name: String,
    pub gallery: Vec<GalleryImage>,
    pub description_heading: String,
    pub description: String,
    pub created_at: Option<DateTime>,
    pub updated_at: Option<DateTime>,
    pub published_at: Option<DateTime>,
    pub sex: Sex,
    pub age: Age,
    pub medical_status: MedicalStatus,
    pub fiv_felv: FivFelv,
    pub castrated: Castrated,
    pub healthy: Healthy,
    pub tags: Vec<String>,
}

macro_rules! attr_enum{
    ( $mod_name:ident, $name:ident, $module:literal, $( $variant:ident ),* ) => {
        pub mod $mod_name {
            pub mod enums {
                use rustler::NifUnitEnum;
                #[derive(Debug, NifUnitEnum)]

                pub enum $name {
                    $(
                        $variant,
                    )*
                }
            }

        }

        #[derive(Debug, NifStruct)]
        #[module = $module]
        pub struct $name{pub value: self::$mod_name::enums::$name}
    }
}

attr_enum![sex, Sex, "Kotkowo.Cat.Sex", Male, Female];

attr_enum![age, Age, "Kotkowo.Cat.Age", Junior, Adult, Senior];

attr_enum![
    medical_status,
    MedicalStatus,
    "Kotkowo.Cat.MedicalStatus",
    TestedAndVaccinated
];

attr_enum![fiv_felv, FivFelv, "Kotkowo.Cat.FivFelv", Positive, Negative];

#[derive(Debug, NifStruct)]
#[module = "Kotkowo.Cat.Castrated"]
pub struct Castrated {
    value: bool,
}

impl From<bool> for Castrated {
    fn from(value: bool) -> Self {
        Self { value }
    }
}

#[derive(Debug, NifStruct)]
#[module = "Kotkowo.Cat.Healthy"]
pub struct Healthy {
    value: bool,
}

impl From<bool> for Healthy {
    fn from(value: bool) -> Self {
        Self { value }
    }
}

File: lib/kotkowo/client/cat.ex
defmodule Kotkowo.Client.Cat do
  @moduledoc false
  alias Kotkowo.Client.Cat.Age
  alias Kotkowo.Client.Cat.Color
  alias Kotkowo.Client.Cat.Sex
  alias Kotkowo.Client.ContactInformation
  alias Kotkowo.Client.Image

  @type castrated() :: boolean()
  @type is_dead() :: boolean()
  @type tags() :: [String.t()]
  @type fiv_felv() :: :negative | :positive
  @type medical_status() :: :tested_and_vaccinated

  defstruct [
    :id,
    :age,
    :name,
    :chip_number,
    :contact_informations,
    :description,
    :color,
    :tags,
    :slug,
    :description_heading,
    :sex,
    :is_dead,
    :fiv_felv,
    :castrated,
    :healthy,
    :medical_status,
    :images
  ]

  @type t() :: %__MODULE__{
          id: String.t() | nil,
          age: Age.t() | nil,
          name: String.t(),
          chip_number: String.t() | nil,
          contact_informations: list(ContactInformation.t()),
          description: String.t() | nil,
          color: Color.t() | nil,
          tags: tags(),
          slug: String.t() | nil,
          description_heading: String.t() | nil,
          sex: Sex.t() | nil,
          is_dead: boolean(),
          fiv_felv: fiv_felv() | nil,
          castrated: boolean(),
          healthy: boolean(),
          medical_status: medical_status() | nil,
          images: [Image.t()]
        }
end
